text/3550-new-range.md

```diff
@@ -0,0 +1,584 @@
+  The iterator for `Range` will yield values from `a` (inclusive) to `b` (exclusive) in steps of one.
+
+- `a..=b` denotes a range from `a` (inclusive) to `b` (inclusive). It resolve to the type `std::ops::range::RangeInclusive`.
+  The iterator for `RangeInclusive` will yield values from `a` (inclusive) to `b` (inclusive) in steps of one.
+
+- `a..` denotes a range from `a` (inclusive) with no upper bound. It resolves to the type `std::ops::range::RangeFrom`.  
+  The iterator for `RangeFrom` will yield values starting with `a` and increasing in steps of one.
+
+These types implement the `IntoIterator` trait, enabling their use directly in a `for` loop:
+```rust
+for n in 0..5 {
+    // `n` = 0, 1, 2, 3, 4
+}
+```
+
+All range types are `Copy` when the bounds are `Copy`, allowing easy reuse:
+```rust
+let range = 0..5;
+
+if a_slice[range].contains(x) {
+    // ...
+}
+if b_slice[range].contains(y) {
+    // ...
+}
+```
+
+For convenience, several commonly-used methods from `Iterator` are present as inherent functions on the range types:
+```rust
+for n in (1..).map(|x| x * 2) {
+    // n = 2, 4, 6, 8, 10, ...
+}
+for n in (0..5).rev() {
+    // n = 4, 3, 2, 1, 0
+}
+```
+
+## Legacy Range Types
+
+In Rust editions prior to 2024, `a..b`, `a..=b`, and `a..` resolved to a different set of types (now found in `std::ops::range::legacy`). These legacy range types did not implement `Copy`, and implemented `Iterator` directly (rather than `IntoIterator`).
+
+This meant that any `Iterator` method could be called on those range types:
+```rust
+let mut range = 0..5;
+assert_eq!(range.next(), Some(0));
+range.for_each(|n| {
+    // n = 1, 2, 3, 4
+});
+```
+
+There exist `From` impls for converting from the new range types to the legacy range types.
+
+### Migrating
+[migrating]: #migrating
+
+In many cases, no changes need to be made at all. This includes most places where a `RangeBounds` or `IntoIterator` is expected:
+```rust
+pub fn takes_range(range: impl std::ops::RangeBounds<usize>) { ... }
+takes_range(0..5); // No changes necessary
+
+pub fn takes_iter(range: impl IntoIterator<usize>) { ... }
+takes_iter(0..5); // No changes necessary
+```
+
+And most places where `Iterator` methods were used directly on the range:
+```rust
+for n in (0..5).rev() { ... } // No changes necessary
+for n in (0..5).map(|x| x * 2) { ... } // No changes necessary
+```
+
+In other cases, `cargo fix --edition 2024` will insert `.into_iter()` as necessary:
+
+```rust
+pub fn takes_iter(range: impl Iterator<usize>) { ... }
+takes_iter((0..5).into_iter()); // Add `.into_iter()`
+
+// Before
+(0..5).for_each(...);
+// After
+(0..5).into_iter().for_each(...); // Add `.into_iter()`
+```
+
+Or fall back to converting to the legacy types:
+
+```rust
+// Before
+pub fn takes_range(range: std::ops::Range<usize>) { ... }
+takes_range(0..5);
+// After
+pub fn takes_range(range: std::ops::range::legacy::Range<usize>) { ... }
+takes_range((0..5).to_legacy());
+```
+
+#### Libraries
```

Comment:
This is part of the "migrating" section, but I think it's very important to point out here that these changes to libraries should happen when any *user* of the library switches to Rust 2024, regardless of whether *the library itself* switches to Rust 2024.

(This makes the edition a non-local choice for libraries with range types in their public interface, which is exactly what worries me.)

Reply:
That's a good point. When I next edit the RFC I will get to make that more clear. Maybe shuffle some things around. I plan on editing the RFC after this interoperability discussion is concluded.